queries/python/textobjects.scm

```diff
@@ -33,3 +32,60 @@
 
 (call) @call.outer
 (call (_) @call.inner)
+
+;; Parameters
+(parameters
+  [(identifier)
+   (tuple)
+   (typed_parameter)
+   (default_parameter)
+   (typed_default_parameter)
+   (list_splat)
+   (dictionary_splat)] @parameter.inner)
+
+(lambda_parameters
+  [(identifier)
+   (tuple)
+   (typed_parameter)
+   (default_parameter)
+   (typed_default_parameter)
+   (list_splat)
+   (dictionary_splat)] @parameter.inner)
+
+(argument_list
+  [(list_splat)
+   (dictionary_splat)
+   (keyword_argument)
+   ; Expressions
+   (comparison_operator)
+   (not_operator)
+   (boolean_operator)
+   (await)
+   (lambda)
+   (conditional_expression)
+   (named_expression)
+   ; Primary expressions
+   (binary_operator)
+   (identifier)
+   (string)
+   (concatenated_string)
+   (integer)
+   (float)
+   (true)
+   (false)
+   (none)
+   (unary_operator)
+   (attribute)
+   (subscript)
+   (call)
+   (list)
+   (list_comprehension)
+   (dictionary)
+   (dictionary_comprehension)
+   (set)
+   (set_comprehension)
+   (tuple)
+   (parenthesized_expression)
+   (generator_expression)
+   (ellipsis)
+  ] @parameter.inner)
```

Comment:
wow, that's too long. I guess it should be ok to be unspecific here. Best would be to add a `type-eq?` predicate and add here a `not-type-eq?` @parameter comment`. See predicates.lua

Reply:
I put this one back to (_), I'll see if I can implement the predicate next week

Reply:
Is that a problem that the capture also captures the comment ? Because those are different nodes, and thus selecting that parameter will just select the very parameter node and not the comment node I guess.

Reply:
Comments are extra nodes, so they can appear anywhere in the grammar

Reply:
Yeah, and that's not an issue, the point I am trying to make here is : is that even a problem to capture a comment node with this very match ?

Reply:
Yeah, there is a problem

```python
    print(
        a + b, # foo
        b + a
    )
```

That will swap the argument with the comment instead of the next argument

Reply:
Problem with line comments is also that they can delete nodes when they move in-front of them:
```
# foo, a+ b 
```

Reply:
Okay then, I agree that there is a problem here then. Now comes my other suggestion about the predicate thing. We should avoid building predicates that are only related to node names and subqueries and instead focus on extending the query syntax to enable that.

So for our given instance, maybe we want to extend the query syntax to allow us to negate some patterns, maybe something like : 
```
(parameters !(comment) @parameter.inner)
```

Or 
```
(parameters (comment)! @parameter.inner)
```

If we face these problems, I guess we are not the only ones, and implementing them is not a problem (but we should be clear on what we would like). So if this kind of solution is the one we would like, then that's fine and we can try to find an implementation for them, and even upstream them if we want !

The goal of queries are to be the most simple and elegant they can be (at least for me). Implementing a predicate that will match the name of a node is not (imo) elegant, while extending the query grammar is. Plus this will allow us to do some crazy awesome things as we go. Please correct me @maxbrunsfeld but I think that adding this to the treesitter query syntax will be both not that hard and beneficial.

Hopefully I does not seem rude, but my point is that we should think of tree-sitter as being part of our project instead of just something that we just use and fight against.

Reply:
Yeah, I think using `(node)!` makes sense, it kind of matches the quantifiers.

Reply:
Another way around would be to allow capturing supertypes

Reply:
I found `(node)!` pretty nice. If we agree on this syntax, I will start working on implementing this.

Reply:
I guess it's simpler with predicates. With a new syntax you always have the problem that it's not clear what it should mean or what other elements it can be combined.

Reply:
@vigoux I agree that the `!` syntax seems elegant, but my concern is that it would be very complex to implement, except for the simple case of a single node like `!(comment)`. For example, it would require a lot of complexity to make it work for complex negated patterns like:

```
(parameters
  !(call
    function: (attribute
      name: (identifier))) @foo)
```

Maybe we could restrict the operator to only allow single nodes, and return a syntax error otherwise. This is somewhat consistent with negated character classes (e.g. `/[^abc]/`) in regular expressions, which only allow characters inside. But is there any way we could change the syntax to make that clearer? The `!` operator sort of makes it appear that it can be composed with arbitrary patterns. Maybe that's ok, but curious if you all have any thoughts.

Reply:
> For example, it would require a lot of complexity to make it work for complex negated patterns like:

Ah, good point. What about `(node!)`, the `!` next to the node, not the group. But maybe `(node)!` + syntax error is ok, and leave it in case it can be implemented in the future for more complex queries?

Reply:
Yeah, I thought about putting the `!` inside of the parentheses, but then we'd need to do something different for negating anonymous nodes like `":"`.

One option is to follow the regular expression syntax, and negate something like this:

```
[^ (comment) ]           ; not a comment
[^ (comment) (string) ]  ; not a comment or string
[^ ";" ":" ]             ; not a colon or semicolon
```

Reply:
Your `[^ ]` proposal is awesome.

One other issue we had is having subqueries, is there any way we can have that, with something like a `__` wildcard (asking as you are here :wink: )?

Reply:
I'm not familiar with the subquery idea. Was there an earlier discussion about it?

Reply:
I guess some discussion happened [here](https://github.com/nvim-treesitter/nvim-treesitter/pull/219#discussion_r459910780).

Reply:
I guess that we should move the problem to a proper `tree-sitter` issue, so that we can have the discussion there and ping the relevant people, @maxbrunsfeld, would you like me to open an issue there ?

Reply:
Yeah, an issue for negation sounds good, and I guess a separate one for the subquery problem.

The subquery issue seems like a more difficult design problem. Taking a step back - why do we need to match all of the nested identifiers inside of an `import_declaration`? I would think you would only want to match the final element (which is class name itself). This could be done via:

```
(import_declaration
  (scoped_identifier
    name: (identifier) @the-imported-class))
```

Reply:
Our problem is that the parsers design is not left recursive for right recursive, and thus we have to go down the whole tree to get the very last identifier.

Reply:
If you're trying to match the last identifier, you can just use the pattern above. A `scoped_identifier` node cannot be right recursive, because its right-hand-side is always just `identifier`.

Reply:
The thin is that we only want to match these if they are a child of a `import_statement` node, thus causing our problem. The problem also occurs for paramater name exctraction is C for example, when you would like the `identifier` child of a `parameter` node.

Reply:
For import statements, the scoped identifier is always the direct child of the import statement right? So I think it should work with the pattern I wrote.

But I can see what you mean about c parameter names, and how that would come up on other situations. We’ll have to do some thinking about how to best solve that.